File: cpp/567.permutation-in-string.cpp
/*
 * @lc app=leetcode id=567 lang=cpp
 *
 * [567] Permutation in String
 */

// @lc code=start
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        // map for storing the frequency of characters in s1 and s2
        map<char, int> m1, m2;

        // iterate through s1 and store the frequency of characters in m1
        for (char c : s1) m1[c]++;

        int n = s1.size();

        for (int i = 0; i < s2.size(); i++) {
            // store the frequency of characters in m2
            m2[s2[i]]++;

            // if window is bigger than string size, remove the first character 
            if (i >= n) {
                // decrement freq of char that is leaving the map
                if (--m2[s2[i - n]] == 0) m2.erase(s2[i - n]);
            }

            // compare the frequency of characters in m1 and m2
            if (m1 == m2) return true;
        }
        // no permutation found
        return false;
    }
};
// @lc code=end

File: python/567.permutation-in-string.py
#
# @lc app=leetcode id=567 lang=python3
#
# [567] Permutation in String
#

# @lc code=start
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        s1_len = len(s1)
        s2_len = len(s2)

        if s1_len > s2_len:
            return False
        
        s1_count = [0] * 26
        s2_count = [0] * 26

        for char in s1: 
            # ord return the unicode of the character
            # ord - ord('a') will give the index of the character in the array
            # increment to count the number of characters
            s1_count[ord(char) - ord('a')] += 1

        for i in range(s2_len):
            s2_count[ord(s2[i]) - ord('a')] += 1
            if i >= s1_len:
                s2_count[ord(s2[i - s1_len]) - ord('a')] -= 1

            if s1_count == s2_count:
                return True
            
        return False
# @lc code=end

